File: 2ndSemester/AlgorithimsAndDataStructuresII/Important_Data_Structures/Linear/java/Stack.java
public class Stack{
  private int[] array_stack;
  private int element_quantity;

  public Stack(){
    this.array_stack = new int[10];
    this.element_quantity = 0;
  }

  public Stack(int size){
    this.array_stack = new int[size];
    this.element_quantity = 0;
  }
  
  public void printStack(){
    for(int i = element_quantity - 1; i >= 0; i--){
      System.out.print("|");
      System.out.print(array_stack[i]);
      System.out.println("|");
    }
    System.out.println("---");
  }

  public void push(int input){
    if(element_quantity == array_stack.length)
      System.out.println("Stack overflow");
    else{
      array_stack[element_quantity++] = input;
      System.out.println(input + " foi inserido");
    }  
  }
  
  public void pop(){
    if(element_quantity > 0){
      System.out.println(array_stack[element_quantity - 1] + " foi removido");
      element_quantity--;
    }
    else 
      System.out.println("Impossivel remover uma pilha vazia");
  }
}

File: 2ndSemester/AlgorithimsAndDataStructuresII/U05_U06-AbstractDataTypes/Dynamic/Stack/C/stack.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Cell{
  int element;
  struct Cell *next;
}Cell;
Cell *top;

void start(){
  top = NULL;
}

Cell *newCell(int x){
  Cell *created = (Cell*)malloc(sizeof(Cell));
  created->element = x;
  created->next = NULL;
  return created;
}

void mostrar(){
  printf("\nYour stack: \n");
  for(Cell *i = top; i != NULL; i = i->next)
    printf("%d ", i->element, ((i->next == NULL)? "\n" : " - "));
}

//inserção
void push(int x){
  Cell *aux = newCell(x);
  aux->next = top;
  top = aux;
  aux = NULL;
}

//remoção
int pop(){
  int removed = 0;
  if(top == NULL)
    printf("Pilha vazia impossivel remover");
  else{
  Cell *aux = top;
  removed = aux->element;
  top = top->next;
  aux->next =NULL;
  aux = NULL;
  }
  return removed;
}

int main(){
  start();
  push(1);
  push(2);
  push(85);
  push(115);
  mostrar();
  pop();
  mostrar();
}
